examples/SHT3X/main/main.cpp

```diff
@@ -0,0 +1,142 @@
+#include <esp_log.h>
+#include <esp_sleep.h>
+#include <freertos/FreeRTOS.h>
+#include <freertos/task.h>
+#include <hulp.h>
+#include <hulp_i2cbb.h>
+
+static const char *TAG = "MAIN";
+
+#define SCL_PIN GPIO_NUM_26
+#define SDA_PIN GPIO_NUM_25
+#define SLAVE_ADDR 0x44     // should be either 0x44 or 0x45
+static gpio_num_t LED_ERR = GPIO_NUM_13;
+static const uint16_t max_temp_difference_raw = 37;     // actually 37.4485 which is about 0.1C
+
+static RTC_SLOW_ATTR ulp_var_t ulp_write_cmd[] = {
+        // write 0x24, 0x00 which is high repeatability with no clock stretching
+        HULP_I2C_CMD_HDR(SLAVE_ADDR, 0x24, 1),
+        HULP_I2C_CMD_1B(0x00),
+};
+static RTC_SLOW_ATTR ulp_var_t ulp_read_cmd[HULP_I2C_CMD_BUF_SIZE(6)] = {
+        HULP_I2C_CMD_HDR_NO_PTR(SLAVE_ADDR, 6),
+};
+static RTC_DATA_ATTR ulp_var_t last_temp;
+
+void init_ulp() {
+    enum {
+        LABEL_I2C_READ,
+        LABEL_I2C_READ_RETURN,
+        LABEL_I2C_WRITE,
+        LABEL_I2C_WRITE_RETURN,
+        LABEL_I2C_ERROR,
+        LABEL_NEGATIVE,
+        LABEL_WAKE,
+    };
+
+    const ulp_insn_t program[] = {
+            // write i2c to request measurement
+            I_MOVO(R1, ulp_write_cmd),
+            M_MOVL(R3, LABEL_I2C_WRITE_RETURN),
+            M_BX(LABEL_I2C_WRITE),
+            M_LABEL(LABEL_I2C_WRITE_RETURN),
+            M_BGE(LABEL_I2C_ERROR, 1),
+
+            // delay 13ms to get result
+            M_DELAY_US_5000_20000(13000),
+
+            // read i2c
+            I_MOVO(R1, ulp_read_cmd),
+            M_MOVL(R3, LABEL_I2C_READ_RETURN),
+            M_BX(LABEL_I2C_READ),
+            M_LABEL(LABEL_I2C_READ_RETURN),
+            M_BGE(LABEL_I2C_ERROR, 1),
+
+            // read first 2 bytes (temperature) to R1
+            I_MOVI(R1,0),
+            I_GET(R1, R1, ulp_read_cmd[HULP_I2C_CMD_DATA_OFFSET]),
+
+            // read last_temp to R2
+            I_MOVI(R2,0),
+            I_GET(R2, R2, last_temp),
+
+            // get difference and store in R0
+            I_SUBR(R0, R1, R2),
+
+            // deal with negative
+            M_BGE(LABEL_NEGATIVE, 0x8000),
+
+            M_BGE(LABEL_WAKE, max_temp_difference_raw),
+            I_HALT(),
+
+            M_LABEL(LABEL_NEGATIVE),
+                M_BL(LABEL_WAKE, 0xFFFF - max_temp_difference_raw),
```

Comment:
Value should be (1 - max_temp_difference_raw) here else the threshold is effectively (max_temp_difference_raw + 2) in the negative direction